plugins/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/FirKotlinUastResolveProviderService.kt

```diff
@@ -35,4 +35,8 @@ interface FirKotlinUastResolveProviderService : BaseKotlinUastResolveProviderSer
     override fun getExpressionType(uExpression: UExpression): PsiType? {
         TODO("Not yet implemented")
     }
+
+    override fun evaluate(uExpression: UExpression): Any? {
+        TODO("Not yet implemented")
```

Comment:
Although we have `IrConstTransformer` here: https://github.com/JetBrains/kotlin/tree/master/compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/evaluate, I don't think it's a good fit here. I.e., we don't want to convert FIR/IR back and forth to evaluate the underlying FIR expression.

FE1.0 UAST uses `BindingContext`'s `COMPILE_TIME_VALUE` for this API. I do remember seeing that before while trying to evaluate constant values of parameters in annotations. That is, I think it's good time to reincarnate FIR-based compile-time constant evaluator: https://github.com/JetBrains/kotlin/pull/3422 cc @mglukhikh

Reply:
If we decide to introduce a FIR-based constant evaluator, I'd be happy to reincarnate that. :D

Maybe too early to ask, but @darthorimar , I think we don't want to use that from `:plugins:uast-kotlin-fir` directly. Does it make sense to introduce, say, `KtCompileTimeConstantProvider` as a separate `KtAnalysisSessionComponent` in `:idea:idea-frontend*` where we can provide an API of form: `KtExpression -> Any?` (or some abstraction of value)?

Reply:
Yeah, adding `KtCompileTimeConstantProvider` to HL API makes sense

Reply:
@jsjeon we discussed the const evaluator internally. Unfortunately, we cannot use an IR evaluator for that purpose because IR can not be built lazily, and we will have to run fir2ir for a rather large chunk of files, probably the whole module. What do you think about introducing a special const evaluator for IDE only? Or maybe we can even reuse the one you've already implemented in https://github.com/JetBrains/kotlin/pull/3422, but only for ide.

Reply:
Let me try reusing a separate FIR module I added in #3422. As per the offline discussion, that one has a simple API entrance that receives `FirExpression` and outputs evaluated constants. On top of it, we can easily add `KtCompileTimeConstantProvider` with `KtExpression` wrapper. I can do that once this one is merged.

Reply:
Btw, why do we need a separate module? The evaluator can be just placed into `idea-frontend-fir` module

Reply:
This is done at https://github.com/JetBrains/kotlin/pull/4451